### ISSUE TITLE
Feature request: Possibility to save information about initial nucleons (spatial+momentum)

### ISSUE THREAD

**Issue by iselyuzh** [smash-transport/smash#5]
Dear SMASH team,

I would like to ask if it would be possible to add an option to SMASH to save in the simulated output an event-by-event information about the time, spatial position, and 4-momentum of individual initial nucleons from each of the colliding nuclei before their 1st collision. In a special case of non-interacting nucleons (spectators) to save the initial spatial position and 4-momentum (with time=0).

This information will allow to connect different geometrical quantities of the fluctuating initial state (e.g. eccentricities) with the final state azimuthal asymmetries (flow harmonics).

Thank you in advance.

Cheers,
Ilya

**Comment by elfnerhannah** [comment]
I think of two options that would realise this and be helpful: 
1) A dump of initial state information on coordinate and momentum space of particles before evolution happens (this applies to all modi of SMASH not only collisions)
2) The option to spill out output at individual intermediate time steps that can be specified in addition to the final time, currently it is only possible to spill out at regular time steps -> this would be useful for the purpose of calculating eccentricities at the overlap or after the passing time of the nuclei for example.

**Comment by stdnmr** [comment]
To extract the information that you want, it is maybe sufficient to add `Only_Final: False` to the output configuration e.g. the default `config.yaml` output-part would then read:
```yaml

Output:
    Output_Interval: 10.0
    Particles:
        Format:          ["Oscar2013"]
        Only_Final:      False

```

This outputs the particles/nucleons at the very beginning of the calculation (when the nuclei are still apart at negative time), at t=0 (this is when the nuclei first touch) and at the output intervals. If you set the `Output_Interval` to the `End_Time` than you get the particles beginning of the calculation, at t=0 and when the calculation stops.

**Comment by iselyuzh** [comment]
If I understand correctly, these options will dump a state of all particle at fixed evolution time, while the interaction of different nucleons can happen at different times (especially this is relevant at low collision energies). From that point of view, I believe it would be good to also have in the output file the 4D (space+time) positions of the nucleons just before their first collision.

**Comment by elfnerhannah** [comment]
So what 

> To extract the information that you want, it is maybe sufficient to add `Only_Final: False` to the output configuration e.g. the default `config.yaml` output-part would then read:
> 
> ```yaml
> Output:
>     Output_Interval: 10.0
>     Particles:
>         Format:          ["Oscar2013"]
>         Only_Final:      False
> ```
> 
> This outputs the particles/nucleons at the very beginning of the calculation (when the nuclei are still apart at negative time), at t=0 (this is when the nuclei first touch) and at the output intervals. If you set the `Output_Interval` to the `End_Time` than you get the particles beginning of the calculation, at t=0 and when the calculation stops.

This would solve the item 1. above, but I think the second one would be still interesting.

**Comment by justinmohs** [comment]
I'm working on an option to specify the output times in terms of a list. That feature should be available in SMASH 1.7.


> If I understand correctly, these options will dump a state of all particle at fixed evolution time, while the interaction of different nucleons can happen at different times (especially this is relevant at low collision energies). From that point of view, I believe it would be good to also have in the output file the 4D (space+time) positions of the nucleons just before their first collision.

This information can be found in the collision output, where the details of the ingoing and outgoing particles of each hadronic interaction are written. I believe the simplest way to extract the time and position of participants just before their first interaction is to enable the extended collision output and identify the first interaction of a particle by checking whether ncoll is 0. The 4-position and 4-momentum can then be read from the same line.

**Comment by stdnmr** [comment]
I am happy to report that the feature to specify the output times will be available in SMASH-1.7.

**Comment by stdnmr** [comment]
Since SMASH-1.7 was released with the requested option, I close the issue. Thanks for the input.
